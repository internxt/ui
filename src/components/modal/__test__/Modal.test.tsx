/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { describe, it, afterEach, expect, vi } from 'vitest';
import { Modal } from '../';
import { afterAll, beforeAll } from 'vitest';

describe('Modal Component', () => {
  const onCloseMock = vi.fn();
  const renderModal = () => {
    return render(
      <Modal isOpen={true} onClose={onCloseMock}>
        <div className="text-center">
          <h2 className="text-lg font-semibold">Title</h2>
          <p className="mt-2">Modal Content</p>

          <button className="mt-4 px-4 py-2 bg-blue-600 text-white rounded hover:bg-blue-700" onClick={onCloseMock}>
            Close
          </button>
        </div>
      </Modal>,
    );
  };

  beforeAll(() => {
    (global as any).IntersectionObserver = class {
      observe() {}
      unobserve() {}
      disconnect() {}
    };
  });

  afterEach(() => {
    vi.clearAllMocks();
    vi.restoreAllMocks();
  });

  afterAll(() => {
    delete (global as any).IntersectionObserver;
  });

  it('should match snapshot when isOpen is true', () => {
    const { container } = renderModal();
    expect(container).toMatchSnapshot();
  });

  it('should render the modal when isOpen is true', () => {
    renderModal();
    expect(screen.getByText('Modal Content')).toBeInTheDocument();
  });

  it('should not render the modal when isOpen is false', () => {
    render(
      <Modal isOpen={false} onClose={onCloseMock}>
        <div>
          Modal Content
          <button>Button</button>
        </div>
      </Modal>,
    );
    expect(screen.queryByText('Modal Content')).not.toBeInTheDocument();
  });

  it('should call onClose when press escape if preventClosing is false', () => {
    renderModal();

    const modalContent = screen.getByTestId('ModalContent');
    expect(modalContent).toBeInTheDocument();

    fireEvent.keyDown(modalContent, { key: 'Escape', code: 'Escape' });

    expect(onCloseMock).toHaveBeenCalledTimes(1);
  });

  it('should not call onClose when press escape if preventClosing is true', () => {
    render(
      <Modal isOpen={true} onClose={onCloseMock} preventClosing={true}>
        <div>
          Modal Content
          <button>Button</button>
        </div>
      </Modal>,
    );

    const modalContent = screen.getByTestId('ModalContent');
    expect(modalContent).toBeInTheDocument();

    fireEvent.keyDown(modalContent, { key: 'Escape', code: 'Escape' });

    expect(onCloseMock).not.toHaveBeenCalled();
  });

  it('should call onClose when click outside if preventClosing is false', () => {
    renderModal();

    const modalContent = screen.getByTestId('ModalContent');
    expect(modalContent).toBeInTheDocument();

    fireEvent.mouseDown(document.body);

    expect(onCloseMock).toHaveBeenCalledTimes(1);
  });

  it('should not call onClose when click outside if preventClosing is true', () => {
    render(
      <Modal isOpen={true} onClose={onCloseMock} preventClosing={true}>
        <div>
          Modal Content
          <button>Button</button>
        </div>
      </Modal>,
    );

    const modalContent = screen.getByTestId('ModalContent');
    expect(modalContent).toBeInTheDocument();

    fireEvent.mouseDown(modalContent);

    expect(onCloseMock).not.toHaveBeenCalled();
  });

  it('should apply custom class names', () => {
    render(
      <Modal isOpen={true} onClose={onCloseMock} className="custom-class">
        Modal Content
        <button>Button</button>
      </Modal>,
    );

    const modalContent = screen.getByTestId('ModalContent');
    expect(modalContent).toHaveClass('custom-class');
  });

  it('should use default width and maxWidth if not provided', () => {
    renderModal();

    const modalContent = screen.getByTestId('ModalContent');
    expect(modalContent).toHaveClass('w-full');
    expect(modalContent).toHaveClass('max-w-lg');
  });

  it('should apply provided width and maxWidth', () => {
    render(
      <Modal isOpen={true} onClose={onCloseMock} width="w-1/2" maxWidth="max-w-xl">
        Modal Content
        <button>Button</button>
      </Modal>,
    );

    const modalContent = screen.getByTestId('ModalContent');
    expect(modalContent).toHaveClass('w-1/2');
    expect(modalContent).toHaveClass('max-w-xl');
  });

  it('should update transition classes after timeout when isOpen is true', async () => {
    renderModal();

    const modalContent = screen.getByTestId('ModalContent');
    expect(modalContent).toHaveClass('opacity-0');
    expect(modalContent).toHaveClass('scale-95');

    await waitFor(() => {
      expect(modalContent).toHaveClass('opacity-100');
      expect(modalContent).toHaveClass('scale-100');
    });
  });

  it('should reset transition classes when isOpen changes to false and closes the modal', async () => {
    const { rerender } = renderModal();

    const modalContent = screen.getByTestId('ModalContent');
    await waitFor(() => {
      expect(modalContent).toHaveClass('opacity-100');
      expect(modalContent).toHaveClass('scale-100');
    });

    rerender(
      <Modal isOpen={false} onClose={onCloseMock}>
        <div>Modal Content</div>
      </Modal>,
    );

    await waitFor(() => {
      expect(modalContent).toHaveClass('opacity-0');
      expect(modalContent).toHaveClass('scale-95');
      expect(screen.queryByText('Modal Content')).not.toBeInTheDocument();
    });
  });
});
