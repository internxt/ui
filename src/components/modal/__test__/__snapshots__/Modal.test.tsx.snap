// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal Component > should match snapshot when isOpen is true 1`] = `
<div>
  <div
    class="m-0"
  >
    <div
      class="
              fixed
              min-h-full
              inset-0
              z-[9999]
              bg-highlight/40
              transition-opacity
              duration-150
              ease-out
              opacity-0
              pointer-events-none           
            "
    />
    <div
      class="
              fixed
              inset-0
              z-[9999]
              flex
              min-h-full
              items-center
              justify-center
              transition-opacity
              duration-150
              ease-out
              overflow-y-auto
              opacity-0
              scale-95
            "
    >
      <section
        class="
                w-full
                max-w-lg
                p-5
                text-gray-100
                rounded-2xl
                bg-surface
                shadow-subtle-hard
                transform
                transition-all
                duration-150
                ease-out
                opacity-0
                scale-95
              "
        data-modal="true"
        data-testid="ModalContent"
      >
        <div
          class="text-center"
        >
          <h2
            class="text-lg font-semibold"
          >
            Title
          </h2>
          <p
            class="mt-2"
          >
            Modal Content
          </p>
          <button
            class="mt-4 px-4 py-2 bg-blue-600 text-white rounded hover:bg-blue-700"
          >
            Close
          </button>
        </div>
      </section>
    </div>
  </div>
</div>
`;
